app/src/main/java/com/guru/bluetooth/MainActivity.kt

```diff
@@ -1,194 +1,248 @@
+        if (bluetoothAdapter?.isEnabled != true) {
+            val enableBtIntent = Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)
+            requestEnableBtLauncher.launch(enableBtIntent)
         }
+    }
 
-        binding.sendButton.setOnClickListener {
-            val fileUri = binding.fileUriTextView.text.toString()
-            if (fileUri.isNotEmpty() && bluetoothDevice != null) {
-                sendFile(getFileUri(fileUri))
-            } else {
-                Toast.makeText(
-                    this,
-                    "Please select a file and connect to a device",
-                    Toast.LENGTH_SHORT
-                ).show()
+    private val receiver = object : BroadcastReceiver() {
+        override fun onReceive(context: Context?, intent: Intent?) {
+            val action: String? = intent?.action
+            when (action) {
+                BluetoothDevice.ACTION_FOUND -> {
+                    val device: BluetoothDevice? =
+                        intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE)
+                    val index = discoveredDevices.indexOf(device)
+                    if (index == -1) {
+                        discoveredDevices.add(device!!)
+                    } else {
+                        discoveredDevices[index] = device!!
+                    }
+                }
             }
         }
-        binding.receiveButton.setOnClickListener {
-            CoroutineScope(Dispatchers.IO).launch {
-                bluetoothFileTransfer.receiveFile()
+    }
+
+    private val nameReceiver = object : BroadcastReceiver() {
+        override fun onReceive(context: Context?, intent: Intent?) {
+            val action: String? = intent?.action
+            when (action) {
+                BluetoothDevice.ACTION_NAME_CHANGED -> {
+                    val device: BluetoothDevice? =
+                        intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE)
+                    val index = discoveredDevices.indexOf(device!!)
+                    if (index == -1) {
+                        discoveredDevices.add(device)
+                    } else {
+                        discoveredDevices[index] = device
+                    }
+                }
             }
         }
     }
 
-    private fun enableBluetooth() {
-        if (bluetoothAdapter == null) {//let
-            Toast.makeText(this, "Device does not support Bluetooth", Toast.LENGTH_SHORT).show()
-            finish()
+    private fun refreshList() {
+        if (!isInScanningMode) {
+            showToast("Please enter the scanning mode first")
+            return
         }
 
-        if (!bluetoothAdapter!!.isEnabled) {
-            val enableBtIntent = Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)
-            requestEnableBtLauncher.launch(enableBtIntent)
+        pairedDevices = bluetoothAdapter?.bondedDevices as Set<BluetoothDevice>
+        val list: ArrayList<BluetoothDevice> = ArrayList()
```

Comment:
can we rename this local variable **list** to something like bluetoothDeviceList